examples/linear_model/plot_poisson_regression_non_normal_loss.py

```diff
@@ -0,0 +1,285 @@
+#
+# 1. Loading datasets and pre-processing
+# --------------------------------------
+#
+# We construct the freMTPL2 dataset by joining the freMTPL2freq table,
+# containing the number of claims (``ClaimNb``) with the freMTPL2sev table
+# containing the claim amount (``ClaimAmount``) for the same policy ids
+# (``IDpol``).
+
+df = load_mtpl2(n_samples=50000)
+
+# Note: filter out claims with zero amount, as the severity model
+# requires strictly positive target values.
+df.loc[(df.ClaimAmount == 0) & (df.ClaimNb >= 1), "ClaimNb"] = 0
+
+# correct for unreasonable observations (that might be data error)
+df["ClaimNb"] = df["ClaimNb"].clip(upper=4)
+df["Exposure"] = df["Exposure"].clip(upper=1)
+
+column_trans = ColumnTransformer(
+    [
+        ("Veh_Driv_Age", KBinsDiscretizer(n_bins=10), ["VehAge", "DrivAge"]),
+        (
+            "Veh_Brand_Gas_Region",
+            OneHotEncoder(),
+            ["VehBrand", "VehPower", "VehGas", "Region", "Area"],
+        ),
+        ("BonusMalus", "passthrough", ["BonusMalus"]),
+        (
+            "Density_log",
+            make_pipeline(
+                FunctionTransformer(np.log, validate=False), StandardScaler()
+            ),
+            ["Density"],
+        ),
+    ],
+    remainder="drop",
+)
+X = column_trans.fit_transform(df)
+
+##############################################################################
+#
+# The number of claims (``ClaimNb``) is a positive integer that can be modeled
+# as a Poisson distribution. It is then assumed to be the number of discrete
+# events occurring with a constant rate in a given time interval
+# (``Exposure``, in units of years). Here we model the frequency
+# ``y = ClaimNb / Exposure``, which is still a (scaled) Poisson distribution,
+# and use ``Exposure`` as `sample_weight`.
+
+df["Frequency"] = df.ClaimNb / df.Exposure
+
+print(
+   pd.cut(df.Frequency, [-1e-6, 1e-6, 1, 2, 3, 4, 5]).value_counts()
+)
+
+print("Average Frequency = {}"
+      .format(np.average(df.Frequency, weights=df.Exposure)))
+
+##############################################################################
+#
+# It worth noting that 96 % of policyholders have zero claims, and if we were
+# to convert this problem into a binary classification task, it would be
+# significantly imbalanced.
+#
+# To evaluate the pertinence of the used metrics, we will consider as a
+# baseline an estimator that returns the mean of the training sample.
+
+df_train, df_test = train_test_split(df, random_state=0)
+
+dummy = make_pipeline(
+    column_trans,
+    DummyRegressor(strategy='mean')
+)
+dummy.fit(df_train, df_train.Frequency,
+          dummyregressor__sample_weight=df_train.Exposure)
+
+##############################################################################
+#
+# The Poisson deviance cannot be computed on negative values predicted by the
+# model, so we set the minimum predicted value to eps,
+
+
+def score_estimator(estimator, df_test, eps=1e-5):
```

Comment:
@rth The cut `eps` to make predictions positive seems quite arbitrary. With the `linreg` model, I get
```python
for i in range(14):
    print("1e-{}: {:.4f}"
          .format(i,
                  mean_poisson_deviance(
                  df_test.Frequency.values, np.fmax(10**(-i), linregr.predict(df_test)),
                  df_test.Exposure.values)))
1e-0: 1.9189
1e-1: 0.5237
1e-2: 0.5180
1e-3: 0.5266
1e-4: 0.5342
1e-5: 0.5416
1e-6: 0.5491
1e-7: 0.5565
1e-8: 0.5640
1e-9: 0.5714
1e-10: 0.5789
1e-11: 0.5863
1e-12: 0.5938
1e-13: 0.6012
```
I then tried a Normal model with log-link `TweedieRegressor(power=0, link="log")`. Oddly, I got better Poisson deviance than with the `PoissonRegressor`.

Reply:
Hmm, yes I agree that's not great. What do you suggest we should do? Removing negativing point from the evaluation or not evaluating altogether if some y values are negatives could be a workaround.

Reply:
I'd say, a model that produces non-positve predictions is not valid. This could also be a business requirement, for example. We could state the invalidity and continue with a Normal with log-link, e.g. `TweedieRegressor(power=0, link='log')`. But I don't know, what to do with the GBT.

Reply:
@rth Also `Ridge` and `GradientBoostingRegressor` produce negative predictions.
My **proposal**:

- Remove `eps` and say instead that a model with negative predictions is invalid.

- Add `lognormal=TweedieRegressor(power=0, link='log')` and increase the `n_samples`  from 50000 to 60000 in `load_mtpl2(n_samples=60000)` such that the Poisson model gets better poisson deviance than the lognormal model.

- Replace the GBRT with `RandomForestRegressor(min_weight_fraction_leaf=1e-2)` as a random forest produces non-negative predictions when trained on non-negative targets.

Reply:
OK, so removed the thresholding on `y_pred` with eps and replaced Gradient Boosting with a Random Forest that does return positive values.

About the first model, the initial motivation was to illustrate the most simple models as something that could be done in a first apporach
 - using `lognormal=TweedieRegressor(power=0, link='log')` is definitely not natural for most scikit-learn users (at least as a baseline model), and I would rather use a existing scikit-learn estimator, so they have something familiar to relate to.
 - `TransformedTargetRegressor` is possible, but then the question is which parametric function to use, why that one and what effect this has, and how this is related to the inverse link function in GLM. I'm not sure I want to go there for this introductory model. BTW, using for instance `np.log1p` and its inverse `np.exp(x) - 1` doesn't guarantee positivity since e.g. `exp(-1) - 1 < 0` unless I'm missing something. A plain `log` also wouldn't work because there are 0 values.

Anyway, for now I would still use a regularized Ridge. It yields around 3% of negative predictions, which if it was just to predict frequency would be fine as one could always clip the predictions at 0. The issue is the Poisson deviance where this wouldn't be valid, there I ignore samples the few samples with negative predictions and mention that score is probably approximate as a result. It's not ideal, but I feel that keeping this example simple is more important.

An alternative could be to return NaN for the Poisson deviance and saying that it is not valid for Ridge.